File: Brands/Podonki/Projects/Content-automation/scripts/github-api-examples.js
/**
 * GitHub API Examples для автоматизации Podonki
 * Использование: node scripts/github-api-examples.js
 */

const fetch = require('node-fetch');

class GitHubAPI {
  constructor() {
    this.token = process.env.GITHUB_TOKEN;
    this.owner = 'YOUR_USERNAME'; // Заменить на реальный юзер
    this.repo = 'podonki-content'; // Заменить на реальный репо
    this.baseUrl = 'https://api.github.com';

    if (!this.token) {
      throw new Error('GITHUB_TOKEN не установлен в переменных окружения');
    }
  }

  async request(method, endpoint, body = null) {
    const url = `${this.baseUrl}${endpoint}`;
    const options = {
      method,
      headers: {
        'Authorization': `token ${this.token}`,
        'Content-Type': 'application/json',
        'Accept': 'application/vnd.github.v3+json',
      },
    };

    if (body) {
      options.body = JSON.stringify(body);
    }

    try {
      const response = await fetch(url, options);

      if (!response.ok) {
        const error = await response.text();
        throw new Error(`GitHub API Error: ${response.status} - ${error}`);
      }

      return await response.json();
    } catch (error) {
      console.error(`❌ Request failed: ${error.message}`);
      throw error;
    }
  }

  // ==================== ISSUES ====================

  /**
   * Создает новую Issue для контент-задачи
   */
  async createContentIssue(title, description, labels = []) {
    console.log(`📝 Creating issue: ${title}`);

    const result = await this.request('POST', `/repos/${this.owner}/${this.repo}/issues`, {
      title,
      body: description,
      labels: ['content', ...labels],
      assignees: [this.owner],
    });

    console.log(`✅ Issue created: #${result.number}`);
    return result;
  }

  /**
   * Создает Issue для каждой идеи из IDEAS-BACKLOG
   */
  async createIssuesFromIdeas(ideas) {
    console.log(`\n📚 Creating ${ideas.length} issues from ideas...`);

    for (const idea of ideas) {
      const description = `
## Идея для контента

**Название:** ${idea.title}
**Тип:** ${idea.type}
**Платформа:** ${idea.platform}

### Описание
${idea.description}

### Теги
${idea.tags.map(tag => `- #${tag}`).join('\n')}

### Критерии успеха
- [ ] Контент соответствует бренду
- [ ] Проверен на качество
- [ ] Готов к публикации
- [ ] Опубликован

### Сроки
Дедлайн: ${new Date(Date.now() + 7 * 24 * 60 * 60 * 1000).toLocaleDateString()}

---
*Создано автоматически GitHub Actions*
      `;

      try {
        await this.createContentIssue(
          `[${idea.type.toUpperCase()}] ${idea.title}`,
          description,
          [idea.platform, idea.type]
        );
      } catch (error) {
        console.error(`⚠️  Failed to create issue for "${idea.title}"`);
      }
    }

    console.log(`✅ All issues created`);
  }

  /**
   * Обновляет Issue (например, добавляет комментарий)
   */
  async addIssueComment(issueNumber, comment) {
    console.log(`💬 Adding comment to issue #${issueNumber}`);

    const result = await this.request(
      'POST',
      `/repos/${this.owner}/${this.repo}/issues/${issueNumber}/comments`,
      { body: comment }
    );

    console.log(`✅ Comment added`);
    return result;
  }

  /**
   * Закрывает Issue
   */
  async closeIssue(issueNumber) {
    console.log(`🚀 Closing issue #${issueNumber}`);

    const result = await this.request(
      'PATCH',
      `/repos/${this.owner}/${this.repo}/issues/${issueNumber}`,
      { state: 'closed' }
    );

    console.log(`✅ Issue closed`);
    return result;
  }

  // ==================== FILES ====================

  /**
   * Читает файл из репо
   */
  async readFile(path) {
    console.log(`📖 Reading file: ${path}`);

    const result = await this.request(
      'GET',
      `/repos/${this.owner}/${this.repo}/contents/${path}`
    );

    // Декодируем base64
    const content = Buffer.from(result.content, 'base64').toString('utf-8');
    return { content, sha: result.sha };
  }

  /**
   * Сохраняет файл в репо
   */
  async writeFile(path, content, message, sha = null) {
    console.log(`💾 Writing file: ${path}`);

    const body = {
      message,
      content: Buffer.from(content).toString('base64'),
      branch: 'main',
    };

    if (sha) {
      body.sha = sha; // Для обновления существующего файла
    }

    const result = await this.request(
      'PUT',
      `/repos/${this.owner}/${this.repo}/contents/${path}`,
      body
    );

    console.log(`✅ File saved: ${path}`);
    return result;
  }

  /**
   * Обновляет JSON файл (читает, изменяет, сохраняет)
   */
  async updateJsonFile(path, updateFn, message) {
    const { content, sha } = await this.readFile(path);

    let data = JSON.parse(content);
    data = updateFn(data);

    const newContent = JSON.stringify(data, null, 2);

    return this.writeFile(path, newContent, message, sha);
  }

  /**
   * Добавляет запись в IDEAS-BACKLOG.json
   */
  async addIdeaToBacklog(idea) {
    console.log(`\n💡 Adding idea to backlog: ${idea.title}`);

    return this.updateJsonFile(
      'data/IDEAS-BACKLOG.json',
      (ideas) => {
        // Добавляем новую идею в начало
        return [
          {
            ...idea,
            created_at: new Date().toISOString(),
            id: `idea-${Date.now()}`,
            quality_score: 0,
            posted: false,
          },
          ...ideas,
        ];
      },
      `Add idea: ${idea.title}`
    );
  }

  // ==================== RELEASES ====================

  /**
   * Создает Release (версия контента)
   */
  async createRelease(tagName, name, body) {
    console.log(`🎯 Creating release: ${name}`);

    const result = await this.request(
      'POST',
      `/repos/${this.owner}/${this.repo}/releases`,
      {
        tag_name: tagName,
        name,
        body,
        draft: false,
        prerelease: false,
      }
    );

    console.log(`✅ Release created: v${result.tag_name}`);
    return result;
  }

  /**
   * Создает Release контента (например, еженедельный выпуск)
   */
  async createContentRelease(weekNumber) {
    const today = new Date();
    const tagName = `content-week-${weekNumber}`;
    const name = `Content Release Week ${weekNumber} (${today.toLocaleDateString()})`;

    const body = `
## Контент для недели ${weekNumber}

**Период:** ${new Date(today.getTime() - today.getDay() * 24 * 60 * 60 * 1000).toLocaleDateString()} - ${today.toLocaleDateString()}

### Статистика
- Идей сгенерировано: TBD
- Постов опубликовано: TBD
- Платформы: TikTok, Telegram, Instagram

### Включено в выпуск
- COMPETITORS-TRACKING.json
- IDEAS-BACKLOG.json
- CONTENT-LOG.json

---
*Версионирование контента Podonki*
    `;

    return this.createRelease(tagName, name, body);
  }

  // ==================== COMMITS ====================

  /**
   * Получает последние коммиты
   */
  async getLatestCommits(limit = 10) {
    console.log(`📋 Getting latest commits...`);

    const result = await this.request(
      'GET',
      `/repos/${this.owner}/${this.repo}/commits?per_page=${limit}`
    );

    return result.map(commit => ({
      sha: commit.sha.substring(0, 7),
      message: commit.commit.message,
      author: commit.commit.author.name,
      date: commit.commit.author.date,
    }));
  }

  /**
   * Получает историю парсинга (коммиты с "parse" в сообщении)
   */
  async getParsingHistory() {
    console.log(`🔍 Getting parsing history...`);

    const commits = await this.getLatestCommits(50);

    return commits.filter(commit =>
      commit.message.toLowerCase().includes('parse') ||
      commit.message.toLowerCase().includes('scrape')
    );
  }

  // ==================== WORKFLOWS ====================

  /**
   * Запускает GitHub Actions workflow вручную
   */
  async triggerWorkflow(workflowName, inputs = {}) {
    console.log(`⚡ Triggering workflow: ${workflowName}`);

    const result = await this.request(
      'POST',
      `/repos/${this.owner}/${this.repo}/actions/workflows/${workflowName}/dispatches`,
      {
        ref: 'main',
        inputs,
      }
    );

    console.log(`✅ Workflow triggered`);
    return result;
  }

  /**
   * Триггерит парсинг вручную
   */
  async triggerCompetitorsParsing(platform = 'all') {
    return this.triggerWorkflow('weekly-competitors-parse.yml', {
      platform,
    });
  }

  /**
   * Триггерит генерацию контента вручную
   */
  async triggerContentGeneration() {
    return this.triggerWorkflow('daily-ideas-generation.yml');
  }

  /**
   * Триггерит публикацию вручную (с выбором индекса)
   */
  async triggerTelegramPosting(ideaIndex = -1) {
    return this.triggerWorkflow('scheduled-telegram-posting.yml', {
      force_post_index: ideaIndex.toString(),
    });
  }
}

// ==================== ПРИМЕРЫ ИСПОЛЬЗОВАНИЯ ====================

async function examples() {
  try {
    const github = new GitHubAPI();

    // Пример 1: Получить последние коммиты
    console.log('\n=== Example 1: Recent commits ===');
    const commits = await github.getLatestCommits(5);
    console.log(commits);

    // Пример 2: Получить историю парсинга
    console.log('\n=== Example 2: Parsing history ===');
    const parsingHistory = await github.getParsingHistory();
    console.log(`Found ${parsingHistory.length} parsing commits`);

    // Пример 3: Прочитать IDEAS-BACKLOG
    console.log('\n=== Example 3: Read IDEAS-BACKLOG ===');
    const { content: ideasContent } = await github.readFile('data/IDEAS-BACKLOG.json');
    const ideas = JSON.parse(ideasContent);
    console.log(`Loaded ${ideas.length} ideas`);

    // Пример 4: Добавить новую идею
    console.log('\n=== Example 4: Add new idea ===');
    const newIdea = {
      title: 'Best vape unboxing tricks via GitHub API',
      description: 'Генерированная идея через GitHub Actions',
      type: 'video',
      platform: 'tiktok',
      tags: ['vape', 'unboxing', 'trends'],
      quality_score: 8,
    };
    await github.addIdeaToBacklog(newIdea);

    // Пример 5: Создать Issue для идеи
    console.log('\n=== Example 5: Create issue ===');
    await github.createContentIssue(
      '[TikTok] Vape unboxing video trend',
      `Идея для видео в тренде\n\n- Целевая аудитория: 16-25\n- Длительность: 15-60 сек`,
      ['tiktok', 'auto-generated']
    );

    // Пример 6: Триггер workflow
    console.log('\n=== Example 6: Trigger workflows ===');
    // await github.triggerContentGeneration();
    // await github.triggerCompetitorsParsing('tiktok');

    console.log('\n✅ All examples completed');

  } catch (error) {
    console.error('❌ Error:', error.message);
  }
}

// Экспорт для использования в других скриптах
module.exports = GitHubAPI;

// Запуск примеров если файл запущен напрямую
if (require.main === module) {
  examples();
}

File: Brands/Podonki/Projects/Content-automation/scripts/github-api-examples.py
"""
GitHub API Examples для автоматизации Podonki (Python версия)
Использование: python scripts/github-api-examples.py
"""

import os
import json
import base64
import requests
from datetime import datetime
from typing import Dict, List, Optional, Any
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class GitHubAPI:
    def __init__(self):
        self.token = os.getenv('GITHUB_TOKEN')
        self.owner = os.getenv('GITHUB_REPO_OWNER', 'YOUR_USERNAME')
        self.repo = os.getenv('GITHUB_REPO_NAME', 'podonki-content')
        self.base_url = 'https://api.github.com'

        if not self.token:
            raise ValueError('GITHUB_TOKEN не установлен')

        self.headers = {
            'Authorization': f'token {self.token}',
            'Content-Type': 'application/json',
            'Accept': 'application/vnd.github.v3+json',
        }

    def _request(self, method: str, endpoint: str, body: Optional[Dict] = None) -> Dict:
        """Выполняет запрос к GitHub API"""
        url = f'{self.base_url}{endpoint}'

        try:
            if method == 'GET':
                response = requests.get(url, headers=self.headers)
            elif method == 'POST':
                response = requests.post(url, headers=self.headers, json=body)
            elif method == 'PUT':
                response = requests.put(url, headers=self.headers, json=body)
            elif method == 'PATCH':
                response = requests.patch(url, headers=self.headers, json=body)
            else:
                raise ValueError(f'Unsupported method: {method}')

            if not response.ok:
                raise Exception(f'GitHub API Error: {response.status_code} - {response.text}')

            return response.json() if response.text else {}

        except Exception as e:
            logger.error(f'❌ Request failed: {str(e)}')
            raise

    # ==================== ISSUES ====================

    def create_content_issue(self, title: str, description: str, labels: List[str] = None) -> Dict:
        """Создает новую Issue для контента"""
        logger.info(f'📝 Creating issue: {title}')

        labels = labels or []
        labels.extend(['content'])

        body = {
            'title': title,
            'body': description,
            'labels': list(set(labels)),
            'assignees': [self.owner],
        }

        result = self._request('POST', f'/repos/{self.owner}/{self.repo}/issues', body)
        logger.info(f'✅ Issue created: #{result.get("number")}')
        return result

    def create_issues_from_ideas(self, ideas: List[Dict]) -> None:
        """Создает Issues для каждой идеи из IDEAS-BACKLOG"""
        logger.info(f'📚 Creating {len(ideas)} issues from ideas...')

        for idea in ideas:
            description = f"""
## Идея для контента

**Название:** {idea.get('title', 'No title')}
**Тип:** {idea.get('type', 'content')}
**Платформа:** {idea.get('platform', 'mixed')}

### Описание
{idea.get('description', 'No description')}

### Теги
{chr(10).join([f'- #{tag}' for tag in idea.get('tags', [])])}

### Критерии успеха
- [ ] Контент соответствует бренду
- [ ] Проверен на качество
- [ ] Готов к публикации
- [ ] Опубликован

### Сроки
Дедлайн: {datetime.now().strftime('%Y-%m-%d')}

---
*Создано автоматически GitHub Actions*
            """

            try:
                self.create_content_issue(
                    f"[{idea.get('type', 'CONTENT').upper()}] {idea.get('title', 'No title')}",
                    description,
                    [idea.get('platform', 'mixed'), idea.get('type', 'content')]
                )
            except Exception as e:
                logger.warning(f'⚠️  Failed to create issue for "{idea.get("title")}": {str(e)}')

        logger.info('✅ All issues created')

    def add_issue_comment(self, issue_number: int, comment: str) -> Dict:
        """Добавляет комментарий к Issue"""
        logger.info(f'💬 Adding comment to issue #{issue_number}')

        body = {'body': comment}
        result = self._request(
            'POST',
            f'/repos/{self.owner}/{self.repo}/issues/{issue_number}/comments',
            body
        )

        logger.info('✅ Comment added')
        return result

    def close_issue(self, issue_number: int) -> Dict:
        """Закрывает Issue"""
        logger.info(f'🚀 Closing issue #{issue_number}')

        body = {'state': 'closed'}
        result = self._request(
            'PATCH',
            f'/repos/{self.owner}/{self.repo}/issues/{issue_number}',
            body
        )

        logger.info('✅ Issue closed')
        return result

    # ==================== FILES ====================

    def read_file(self, path: str) -> tuple[str, str]:
        """Читает файл из репо"""
        logger.info(f'📖 Reading file: {path}')

        result = self._request('GET', f'/repos/{self.owner}/{self.repo}/contents/{path}')

        # Декодируем base64
        content = base64.b64decode(result['content']).decode('utf-8')
        sha = result['sha']

        return content, sha

    def write_file(self, path: str, content: str, message: str, sha: Optional[str] = None) -> Dict:
        """Сохраняет файл в репо"""
        logger.info(f'💾 Writing file: {path}')

        encoded_content = base64.b64encode(content.encode('utf-8')).decode('utf-8')

        body = {
            'message': message,
            'content': encoded_content,
            'branch': 'main',
        }

        if sha:
            body['sha'] = sha

        result = self._request('PUT', f'/repos/{self.owner}/{self.repo}/contents/{path}', body)
        logger.info(f'✅ File saved: {path}')
        return result

    def update_json_file(self, path: str, update_fn, message: str) -> Dict:
        """Обновляет JSON файл"""
        content, sha = self.read_file(path)

        try:
            data = json.loads(content)
        except json.JSONDecodeError:
            data = []

        data = update_fn(data)
        new_content = json.dumps(data, indent=2, ensure_ascii=False)

        return self.write_file(path, new_content, message, sha)

    def add_idea_to_backlog(self, idea: Dict) -> Dict:
        """Добавляет идею в IDEAS-BACKLOG.json"""
        logger.info(f'💡 Adding idea to backlog: {idea.get("title")}')

        def update_fn(ideas):
            new_idea = {
                **idea,
                'created_at': datetime.now().isoformat(),
                'id': f'idea-{int(datetime.now().timestamp() * 1000)}',
                'quality_score': 0,
                'posted': False,
            }
            return [new_idea] + ideas

        return self.update_json_file(
            'data/IDEAS-BACKLOG.json',
            update_fn,
            f'Add idea: {idea.get("title")}'
        )

    # ==================== RELEASES ====================

    def create_release(self, tag_name: str, name: str, body: str) -> Dict:
        """Создает Release"""
        logger.info(f'🎯 Creating release: {name}')

        release_body = {
            'tag_name': tag_name,
            'name': name,
            'body': body,
            'draft': False,
            'prerelease': False,
        }

        result = self._request('POST', f'/repos/{self.owner}/{self.repo}/releases', release_body)
        logger.info(f'✅ Release created: v{result.get("tag_name")}')
        return result

    def create_content_release(self, week_number: int) -> Dict:
        """Создает Release контента за неделю"""
        today = datetime.now()
        tag_name = f'content-week-{week_number}'
        name = f'Content Release Week {week_number} ({today.strftime("%Y-%m-%d")})'

        body = f"""
## Контент для недели {week_number}

**Период:** {today.strftime("%Y-%m-%d")}

### Статистика
- Идей сгенерировано: TBD
- Постов опубликовано: TBD
- Платформы: TikTok, Telegram, Instagram

### Включено в выпуск
- COMPETITORS-TRACKING.json
- IDEAS-BACKLOG.json
- CONTENT-LOG.json

---
*Версионирование контента Podonki*
        """

        return self.create_release(tag_name, name, body)

    # ==================== COMMITS ====================

    def get_latest_commits(self, limit: int = 10) -> List[Dict]:
        """Получает последние коммиты"""
        logger.info('📋 Getting latest commits...')

        result = self._request(
            'GET',
            f'/repos/{self.owner}/{self.repo}/commits?per_page={limit}'
        )

        return [
            {
                'sha': commit['sha'][:7],
                'message': commit['commit']['message'],
                'author': commit['commit']['author']['name'],
                'date': commit['commit']['author']['date'],
            }
            for commit in result
        ]

    def get_parsing_history(self) -> List[Dict]:
        """Получает историю парсинга (коммиты с "parse" в сообщении)"""
        logger.info('🔍 Getting parsing history...')

        commits = self.get_latest_commits(50)
        return [
            c for c in commits
            if 'parse' in c['message'].lower() or 'scrape' in c['message'].lower()
        ]

    # ==================== WORKFLOWS ====================

    def trigger_workflow(self, workflow_name: str, inputs: Dict = None) -> Dict:
        """Запускает GitHub Actions workflow вручную"""
        logger.info(f'⚡ Triggering workflow: {workflow_name}')

        body = {
            'ref': 'main',
            'inputs': inputs or {},
        }

        result = self._request(
            'POST',
            f'/repos/{self.owner}/{self.repo}/actions/workflows/{workflow_name}/dispatches',
            body
        )

        logger.info('✅ Workflow triggered')
        return result

    def trigger_competitors_parsing(self, platform: str = 'all') -> Dict:
        """Триггерит парсинг конкурентов"""
        return self.trigger_workflow('weekly-competitors-parse.yml', {'platform': platform})

    def trigger_content_generation(self) -> Dict:
        """Триггерит генерацию контента"""
        return self.trigger_workflow('daily-ideas-generation.yml')

    def trigger_telegram_posting(self, idea_index: int = -1) -> Dict:
        """Триггерит публикацию в Telegram"""
        return self.trigger_workflow(
            'scheduled-telegram-posting.yml',
            {'force_post_index': str(idea_index)}
        )


# ==================== ПРИМЕРЫ ИСПОЛЬЗОВАНИЯ ====================

async def examples():
    """Примеры использования GitHub API"""
    try:
        github = GitHubAPI()

        # Пример 1: Получить последние коммиты
        print('\n=== Example 1: Recent commits ===')
        commits = github.get_latest_commits(5)
        for commit in commits:
            print(f"{commit['sha']} - {commit['message']} ({commit['author']})")

        # Пример 2: Получить историю парсинга
        print('\n=== Example 2: Parsing history ===')
        parsing_history = github.get_parsing_history()
        print(f'Found {len(parsing_history)} parsing commits')
        for commit in parsing_history[:3]:
            print(f"  - {commit['sha']}: {commit['message']}")

        # Пример 3: Прочитать IDEAS-BACKLOG
        print('\n=== Example 3: Read IDEAS-BACKLOG ===')
        ideas_content, _ = github.read_file('data/IDEAS-BACKLOG.json')
        ideas = json.loads(ideas_content)
        print(f'Loaded {len(ideas)} ideas')

        # Пример 4: Добавить новую идею
        print('\n=== Example 4: Add new idea ===')
        new_idea = {
            'title': 'Best vape unboxing tricks via GitHub API',
            'description': 'Генерированная идея через GitHub Actions',
            'type': 'video',
            'platform': 'tiktok',
            'tags': ['vape', 'unboxing', 'trends'],
            'quality_score': 8,
        }
        github.add_idea_to_backlog(new_idea)

        # Пример 5: Создать Issue для идеи
        print('\n=== Example 5: Create issue ===')
        github.create_content_issue(
            '[TikTok] Vape unboxing video trend',
            'Идея для видео в тренде\n\n- Целевая аудитория: 16-25\n- Длительность: 15-60 сек',
            ['tiktok', 'auto-generated']
        )

        # Пример 6: Триггер workflow
        print('\n=== Example 6: Trigger workflows ===')
        # Раскомментировать для реального использования:
        # github.trigger_content_generation()
        # github.trigger_competitors_parsing('tiktok')

        print('\n✅ All examples completed')

    except Exception as e:
        logger.error(f'❌ Error: {str(e)}')


if __name__ == '__main__':
    import asyncio
    asyncio.run(examples())
